ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,308 @@
+                                    std::vector<std::string>& input_subscripts,
+                                    std::string& output_subscript)
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_parse_equation);
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = equation.find("->");
+    auto input_subscripts_str = equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    bool is_ellipsis_met = false;
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        bool local_is_ellipsis_met = false;
+        // check that input subscript contains only alphabetic letter or ellipsis
+        NGRAPH_CHECK(is_subscript_correct(input_subscript, local_is_ellipsis_met),
+                     "Input subscript of Einsum equation must consist of either only "
+                     "alphabetic letters or alphabetic letters with one ellipsis.");
+
+        // mark that ellipsis is met at least in one input subscript
+        if (local_is_ellipsis_met)
+        {
+            is_ellipsis_met = true;
+        }
+        input_subscripts.push_back(input_subscript);
+    }
+
+    if (pos_output_delimeter == std::string::npos)
+    {
+        // recover output subscript
+        output_subscript = "";
+        for (auto const& input_subscript : input_subscripts)
+        {
+            for (auto const& label : input_subscript)
+            {
+                if (std::isalpha(label) && output_subscript.find(label) == std::string::npos)
+                {
+                    output_subscript += label;
+                }
+            }
+        }
+        std::sort(output_subscript.begin(), output_subscript.end());
+        if (is_ellipsis_met)
+        {
+            output_subscript = "..." + output_subscript;
+        }
+    }
+    else
+    {
+        output_subscript = equation.substr(pos_output_delimeter + 2);
+        bool output_is_ellipsis_met = false;
+
+        // check that the output subscript has the correct format
+        NGRAPH_CHECK(is_subscript_correct(output_subscript, output_is_ellipsis_met),
+                     "Output subscript of Einsum equation must consist of either only "
+                     "alphabetic letters or alphabetic letters with one ellipsis.");
+
+        // if the ellipsis is met in input subscripts, one ellipsis must be in the output subscript
+        NGRAPH_CHECK(is_ellipsis_met == output_is_ellipsis_met,
+                     "Output subscript of Einsum equation must contain one ellipsis if "
+                     "ellipsis is met in any input subscript.");
+    }
+}
+
+std::vector<std::string> op::v7::Einsum::extract_labels(const std::string& subscript)
+{
+    NGRAPH_OP_SCOPE(v7_Einsum_extract_labels);
+
+    std::vector<std::string> labels;
```

Comment:
I think you should reserve vector. for example vector.reserve(subscript_length)

Reply:
subscript has relatively small size so it makes sense in case of huge size containers